sympy/geometry/ellipse.py

```diff
@@ -695,27 +692,64 @@ def is_tangent(self, o):
+                         (self.bounds[1] == o.bounds[1] and self.bounds[3] == self.bounds[3])):
+                return True
+            else:
+                return False
+        elif isinstance(o, Line2D):
+            if len(self.intersection(o)) == 1:
+                return True
+            else:
                 return False
-            return (inter is not None and len(inter) == 1
-                    and isinstance(inter[0], Point))
-        elif isinstance(o, LinearEntity):
-            inter = self.intersection(o)
-            if inter is not None and len(inter) == 1:
-                return inter[0] in o
+        elif isinstance(o, Ray2D):
+            if len(self.intersection(o)) == 1 and not any([self.encloses_point(point) for point in o.points]):
+                return True
             else:
                 return False
-        elif isinstance(o, Polygon):
-            c = 0
-            for seg in o.sides:
-                inter = self.intersection(seg)
-                c += len([True for point in inter if point in seg])
-            return c == 1
+        elif isinstance(o, (Segment2D, Polygon)):
+            segments = o.sides if isinstance(o, Polygon) else [o]
+            any_intersect = False
+            for side in segments:
+                intersect = self.intersection(side)
+                ends_of_segment = side.points
+                all([(self.encloses_point(vertex) or vertex in self) for vertex in ends_of_segment])
+                if len(intersect) == 1 and \
+                        all([(self.encloses_point(point) or point in self) for point in ends_of_segment]):
+                    any_intersect = True
+                    continue
+                elif len(intersect) == 1 and \
+                        all([(not self.encloses_point(point) or point in self) for point in ends_of_segment]):
+                    any_intersect = True
+                    continue
+                elif len(intersect) == 2 and (ends_of_segment[0] in self and ends_of_segment[1] in self):
+                    any_intersect = True
+                    continue
+                elif len(intersect) == 0:
+                    continue
+                else:
+                    return False
+            if isinstance(o, Polygon):
+                if all([(self.encloses_point(vertex) or vertex in self) for vertex in o.vertices]) or \
+                        not any([self.encloses_point(vertex) or vertex in self for vertex in o.vertices]):
+                    return any_intersect
+                else:
+                    return False
+            else:
+                return any_intersect
+        elif isinstance(o, (LinearEntity3D, Point3D)):
+            raise TypeError('Entity must be two dimensional, not three dimensional')
+
         else:
-            raise NotImplementedError("Unknown argument type")
+            raise TypeError('Wrong type of argument were put')
```

Comment:
I wouldn't change this: if the object has a surface or boundary of any kind we may not have handled it so the NotImplemented error let's the user know that maybe it *could* be calculated but SymPy has not implemented that method.